<a href="https://colab.research.google.com/github/AdamStamm/bootcamp_alura/blob/main/Notebooks/Bootcamp_Data_Science_Modulo_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Apresentação

Esse notebook contempla os códigos aprendidos no **Módulo 3** do Bootcamp Data Science aplicada desenvolvido pela [Alura](http://alura.com.br). O foco de aprendizado é **Análise descritiva e testes estatíticos**.

#Importando bibliotecas.

In [24]:
import pandas as pd
from zipfile import ZipFile
 
#biblioteca de fluxo de entrada e saída
#importamos o pacote de leitura binária
from io import BytesIO
import requests

- A biblioteca IO foi feita para controle de fluxos de entrada/saída de dados, no módulo anterior usamos o pacote StringIO da io para capturar o fluxo de dados que estavam na memória (copy and paste). Agora usaremos para capturar o fluxo de dados binários.
Mais informações: https://docs.python.org/3/library/io.html
 
- A biblioteca requests permite que você faça requisições HTTP de maneira fácil sem a necessidade de uma API.
mais informações [aqui](https://requests.readthedocs.io/pt_BR/latest/user/quickstart.html) ou [aqui](https://www.digitalocean.com/community/tutorials/how-to-get-started-with-the-requests-library-in-python-pt)

#Aula 1 - Educação, saúde e tipos de dados.

##Carregando dados

Dessa vez carregaremos os dados de um arquivo zipado, sem descompacta-lo no computador, por maior comodidade upei o arquivo zip direto no meu [Github](https://github.com/AdamStamm/bootcamp_alura/).

In [25]:
#Fazendo requisição binária do arquivo zip do github
path = 'https://github.com/AdamStamm/bootcamp_alura/blob/main/dados/PeNSE_2015_AMOSTRA2.zip?raw=true'
r = requests.get(path)
zip_file = BytesIO(r.content)
 
# Leitura do arquivo zip
zip = ZipFile(zip_file) 
 
#pega o caminho da lista de arquivos
caminho_arquivo_zip = zip.namelist()[1]
 
df = pd.read_csv(zip.open(caminho_arquivo_zip), sep=';')

In [26]:
df.head()

ANOPESQ  PAIS  REGEOGR  VB00004  VB01001  ...  V0008  V0041  aluno  escola  turma
0     2015    76        1        1        2  ...      4      1      1       1    296
1     2015    76        1        1        2  ...      4      1      2       1    296
2     2015    76        1        1        1  ...      4      1      3       1    296
3     2015    76        1        1        1  ...      4      1      4       1    296
4     2015    76        1        1        1  ...      4      1      5       1    296

[5 rows x 181 columns]

Algumas propriedades do ZipFile

In [27]:
#mostrando o objeto que o ZipFile possui.  
zip

<zipfile.ZipFile file=<_io.BytesIO object at 0x7f259f86cba0> mode='r'>

In [28]:
#vendo a lista de arquivos contidos no .zip
zip.infolist()

[<ZipInfo filename='arquivos csv/' external_attr=0x10>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV' compress_type=deflate external_attr=0x20 file_size=7262170 compress_size=1235667>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV' compress_type=deflate external_attr=0x20 file_size=11420533 compress_size=1329882>,
 <ZipInfo filename='arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV' compress_type=deflate external_attr=0x20 file_size=5000642 compress_size=114324>,
 <ZipInfo filename='arquivos sas/' external_attr=0x10>,
 <ZipInfo filename='arquivos sas/pense_amostra2_aluno.sas7bdat' compress_type=deflate external_attr=0x20 file_size=24691712 compress_size=1957530>,
 <ZipInfo filename='arquivos sas/pense_amostra2_alunoescola.sas7bdat' compress_type=deflate external_attr=0x20 file_size=45253632 compress_size=2461932>,
 <ZipInfo filename='arquivos sas/pense_amostra2_escola.sas7bdat' compress_type=deflate external_attr=0x20 file_size=18105344 compress_size=332162>,
 <Zip

In [29]:
#outra forma de ver os arquivos, retornando uma lista
zip.namelist()

['arquivos csv/',
 'arquivos csv/PENSE_AMOSTRA2_ALUNO.CSV',
 'arquivos csv/PENSE_AMOSTRA2_ALUNOESCOLA.CSV',
 'arquivos csv/PENSE_AMOSTRA2_ESCOLA.CSV',
 'arquivos sas/',
 'arquivos sas/pense_amostra2_aluno.sas7bdat',
 'arquivos sas/pense_amostra2_alunoescola.sas7bdat',
 'arquivos sas/pense_amostra2_escola.sas7bdat',
 'arquivos xlsx/',
 'arquivos xlsx/PENSE_AMOSTRA2_ALUNO.xlsx',
 'arquivos xlsx/PENSE_AMOSTRA2_ALUNOESCOLA.xlsx',
 'arquivos xlsx/PENSE_AMOSTRA2_ESCOLA.xlsx',
 'Dicionario_PENSE_Microdados_AMOSTRA2.xls']

Propriedade do requests

In [30]:
#verificando a URL que a requests capturou
r.url

'https://raw.githubusercontent.com/AdamStamm/bootcamp_alura/main/dados/PeNSE_2015_AMOSTRA2.zip'

Iniciando análises iniciais

In [39]:
#verificando o tamanho do data frame
print(f'O conjunto tem {df.shape[0]} linhas e {df.shape[1]} colunas')

O conjunto tem 16556 linhas e 181 colunas


In [44]:
#visualizando colunas
df.columns

Index(['ANOPESQ', 'PAIS', 'REGEOGR', 'VB00004', 'VB01001', 'VB01002',
       'VB01003', 'VB01004', 'VB01005', 'VB01006',
       ...
       'ESTRATO_EXP', 'ESTRATOGEOREG', 'PESO', 'V0006', 'V0007', 'V0008',
       'V0041', 'aluno', 'escola', 'turma'],
      dtype='object', length=181)

Percebemos que existem colunas com os nomes das variáveis difíceis de serem identificados, por essa razão é necessário estudarmos o dicionário de variáveis que acompanha o arquivo zip. Uma outra forma de abordagem é observar a forma desses dados como faremos a seguir.

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16556 entries, 0 to 16555
Columns: 181 entries, ANOPESQ to turma
dtypes: float64(7), int64(171), object(3)
memory usage: 22.9+ MB


Percebemos que existem 7 colunas do tipo float, 171 do tipo inteiro e 3 objects.

In [47]:
#visualizando dados por tipo
df.select_dtypes(include='object')

VB17003 VB17004        PESO
0        37,1     157  299,735235
1        50,2     152  355,170081
2        48,2   161,4  299,735235
3        69,4   172,5  355,170081
4        78,9   167,5  299,735235
...       ...     ...         ...
16551    56,7   167,9  447,815539
16552    61,9   185,8   592,50744
16553    60,9   159,4  401,131512
16554    47,3   157,8  442,718091
16555    62,6   163,9   592,50744

[16556 rows x 3 columns]

Note que o tipo object em geral trata de strings, mas temos 3 colunas de tipo float sendo tratadas como strings, isso acontece por que no nosso idioma pontos flutuantes são debitados por vírgula, no formato americano são demorados por ponto, isso faz com que o código seja classificado como string na hora do carregamento dos dados. Para corrigir isso basta, na hora do carregamento dos dados, informar essa característica incluindo os parâmetros thousands para milhar e decimal para decimais.
 
Também é importante olhar o dicionário de dados, se verificarmos a coluna PESO refere-se ao peso *estatístico* dos dados e **não** do peso dos alunos.

#Desafio 1: Descompactar a pasta zipada utilizando outro método.

Para esse desafio irei fazer upload da pasta para o Google drive uma vez que o GitHub não permite a manipulação de dados sem que sejam dados commits. Portanto irei montar o drive e então trabalhar com o zip.
 
Após o drive montado irei utilizar comandos de Shell Script, ou seja, comandos bashrc. Os servidores colab utilizam por baixo dos panos o sistema Unix e isso permite esse tipo de utilização. Para enviarmos comandos para o sistema usamos o sinal de exclamação antes do comando.

In [31]:
#deixarei a celula comentada pois ainda irei trabalhar muito com esse notebook
#!unzip /content/drive/MyDrive/Dados_Bootcamp/PeNSE_2015_AMOSTRA2.zip -d /content/drive/MyDrive/Dados_Bootcamp

In [32]:
#visualizando arquivos extraídos
!ls /content/drive/MyDrive/Dados_Bootcamp

'arquivos csv'	  Dicionario_PENSE_Microdados_AMOSTRA2.xls
'arquivos sas'	  PeNSE_2015_AMOSTRA2.zip
'arquivos xlsx'


#Desafio 2: Estudar a base PeNSE.

In [33]:
#incluindo vídeo no notebook
from IPython.core.display import HTML
 
HTML('<iframe id="ytplayer" type="texto/html" width="500" height="300" \ src="https://www.youtube.com/embed/watch?v=z4BBCHZUvD4" frameborder="2"/>')

Os dados que serão trabalhados são da Pesquisa Nacional de Saúde do Escolar de 2015. Essa foi a terceira edição da PeNSE e foi a primeira que inclui alunos do ensino médio na faixa dos 13 aos 17 anos, as edições anteriores contemplavam apenas alunos até o nono ano (antiga oitava série).
 
Os estudantes são selecionados de um conjunto de escolas previamente definido (Amostra), e para cada um desses grupos escolares foi criado um plano tabular.
 
A Pesquisa é realizada 
graças à parceria do Ministério da Saúde com o Instituto Brasileiro de Geografia e Estatística e o apoio do Ministério da 
Educação. Ela faz parte da Vigilância dos Fatores de Risco e Proteção das Doenças Crônicas do Brasil. Em 2015, com a inclusão da amostra de estudantes de 13 a 17 anos, tornou a PeNSE 
comparável aos inquéritos internacionais.